### ISSUE TITLE
New `mocha-no-side-effect-code` rule

### ISSUE THREAD

**Issue by loicraux** [Microsoft/tslint-microsoft-contrib#85]
**tl;dr**: This rule to ensure no assignment statements or function calls in tests outside mocha `before()`, `after()`, `beforeEach()`, `afterEach()` and `it()` scopes.

**Reason**: You cannot predict (or you are not supposed to know as a developer writing tests) when code containing assignment statements or function calls in tests outside mocha `before()`, `after()`, `beforeEach()`, `afterEach()` and `it()` is indeed executed (this is up to requirejs, order of requires for the tests files and also up to Mocha for execution of describe().
This leads to some tests having side effects to some other tests.

With this rule, only variable or function declarations would be allowed outside the scopes listed above.
(Also `import`s would be allowed)

**Comment by HamletDRC** [comment]
@s-timofte @loicraux I ran this rule on /common and it found some interesting stuff. I think we could enable this rule fairly easily and I think the cleanup of the test cases would benefit us. 

I configured the rule with: 

    "mocha-no-side-effect-code": [true, { ignore: '^(RestDataFactory\\.create|sinon\.spy).*' }],

**Comment by johnemau** [comment]
What is the guidance for writing [dynamically generated tests](https://mochajs.org/#dynamically-generating-tests) and enabling this rule?

**Comment by kkor** [comment]
Same question here, according to Mochas documentation you can generate tests dynamically like this:
https://mochajs.org/#dynamically-generating-tests

But this will trigger mocha-no-side-effect-code.

**Comment by JoshuaKGoldberg** [comment]
@johnemau @kkor good question. It's pretty difficult to tell statically whether side effect code is legitimately setting up dynamic tests or otherwise doing dangerous work. If you have a proposal for how we could determine it, it'd be great if you could file an issue with details.

Until then your best bet is to just disable the rule, either in your `tslint.json` or `with `// tslint:disable:mocha-no-side-effect-code` lines.

**Comment by WillSquire** [comment]
The thing with Mocha and Jest's way of preparing tests with `before()`, `beforeAll()`, etc, is that it involves using `var` or `let` to perform variable reassignment. If the idea is to be more functional and avoid side effects, then this rule seems to be a bit of a contradiction to itself?
